models/image.py

```diff
@@ -487,8 +471,44 @@ def init_on_load(self):
         if this_object_session is not None:  # if just loaded, should usually have a session!
             self.load_upstream_products(this_object_session)
 
+    def merge_all(self, session):
+        """Use safe_merge to merge all the downstream products and assign them back to self.
+
+        This includes: sources, psf, wcs, zp.
+        This will also merge relationships, such as exposure or upstream_images,
+        but that happens automatically using SQLA's magic.
+
+        Must provide a session to merge into. Need to commit at the end.
+
+        Returns the merged image with all its products on the same session.
+        """
+        new_image = self.safe_merge(session=session)
+        session.flush()
+        if self.sources is not None:
+            self.sources.image = new_image
+            self.sources.image_id = new_image.id
+            self.sources.provenance_id = self.sources.provenance.id if self.sources.provenance is not None else None
+            new_image.sources = self.sources.merge_all(session=session)
+            new_image.wcs = new_image.sources.wcs
+            new_image.zp = new_image.sources.zp
+            new_image.cutouts = new_image.sources.cutouts
```

Comment:
I think this subverts lazy loading.  

I think if self.cutouts is None, I don't think you want to set new_image.cutouts here.  (Same for other fields.)

Reply:
This is supposed to only set the attributes that were already there. 
You cannot lazy load cutouts or sources without specifying the provenance (which means it never gets lazy loaded, only explicitly loaded). 
In this case you have to do this for the merge to work. It essentially replaces the "recursive merge".